[git-webkit] Allow for re-running pre-pr checks

JonWBedard opened the pull request:
#### 052c4c37d40f04042013f66db2bd09c63efdce28
<pre>
[git-webkit] Allow for re-running pre-pr checks
<a href="https://bugs.webkit.org/show_bug.cgi?id=242536">https://bugs.webkit.org/show_bug.cgi?id=242536</a>
rdar://97093963

Reviewed by Dewei Zhu.

In webkit-patch, contributors could modify their patch after the style-checker
ran but before the change was uploaded. git-webkit should support a similar
functionality when running pre-pr checks.

* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.pre_pr_checks): If a pre-pr check fails, offer to re-run the
check and if possible, add any modified files to the most recent commit.
(PullRequest.create_pull_request): Pass --no-add onto pre_pr_checks.

Canonical link: <a href="https://commits.webkit.org/270007@main">https://commits.webkit.org/270007@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8cd1569712ceadc608b9631400210fa61b77a6f5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24266 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/2374 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/25347 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26397 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/22337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/24534 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/4011 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24731 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24509 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/1895 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20967 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26986 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/24436 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/1641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21888 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/28101 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/22125 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22191 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/25894 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1576 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/19225 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/24037 "Passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/2854 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1975 "Built successfully") | | | 
| [  ~~🛠 🧪 unsafe-merge~~](https://ews-build.webkit.org/#/builders/22/builds/3098 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1942 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
```diff
@@ -316,18 +316,30 @@ def pre_pr_checks(cls, repository):
+                if add_edits and run(
+                    [repository.executable(), 'commit', '-a', '--amend', '--date=now', '--no-edit'],
+                    cwd=repository.root_path,
+                ).returncode:
```
I understand the use case, but committing during what is effectively a `git push` is way too magical for my liking. This change makes our workflow more like subversion and less like git, which IMO is the wrong direction to move in. 

With `-a`, I expect this to lead to unexpected changes being pushed (with the associated security/NPS concerns). It seems like it would also amend the wrong commit if you're pushing a multi-commit PR.

What I would vastly prefer is moving style checking to a pre-commit hook, where it's much less hassle to ask a developer to fix their style bug, re-add, and then restart the commit. I don't think we would need auto-rerun in pre-commit, but maybe there is an argument to have it for folks who prefer git-webkit to match webkit-patch as much as possible, and use `git-webkit pr` to make their commit.

JonWBedard replied:
I think the big piece this comment misses is that many developers _want_ changes added when `git-webkit pr` is run (and while I'll let @smfr speak for himself, I believe the workflow here uses `git-webkit pr` without any commit and just modified files, so a pre-commit hooks doesn't really help).

I think your feedback is good, though, and it may be that the current design of this PR is overly-eager to add changes. In this iteration, adding files is powered by `--add` and `--no-add`, so it should do the same thing that `git-webkit pr` already does (although `-A` will add new files too, which is a problem and should be changed).

I think we should have a mode with `git-webkit` that does add style modifications, but I can see arguments that this should be configured by repository settings, would be interested to hear @smfr, @hortont424 and @marcoscaceres (who are involved in the motivating bug) what thing

smfr replied:
My primary request is to have the style checker run before `git webkit pr` makes the branch and commits. This feels more efficient than "let it run later and then re-run PR stuff". If someone has already committed and asks `git webkit pr` to not commit, then running it later seems fine.

JonWBedard replied:
There is an interesting edge case, which is the reason this check is where it currently is: `check-webkit-style` won't check style of a new, but uncommitted, file. I think this means that we would basically have to run the style checker twice if we ran it pre-commit.

My understanding is that your primary concern, @smfr, is that you don't want to have to break out of `git-webkit` and add files manually after failing the style checker, right?

aj062 replied:
+1 for @smfr comment above. 
Currently, if `git-webkit pr` finds any style error, the process to fix that and ensuring that those fixes become part of the pr being created is so non-intuitive/long.

JonWBedard replied:
I agree, which was/is the point of this PR.

To be clear about what the behavior change proposed in this PR is: When a pre-pr check fails (at the moment, our pre-PR check is just the style checker, but I know some contributors make diff displaying a pre-PR check too), this PR will ask the user if they would like to continue uploading their PR, giving them 3 options:
- `Continue`: Ignore the failed check and upload the PR anyways
- `Re-Run`: If `git-webkit pr` is configured to automatically add and commit files, look for any changes and add them to the previous commit before re-running the failed check
- `No`: Exit the program.

My understanding of the frustration here is that the `Re-Run` option would basically fix them. In the event contributors prefer managing their own `git` state (like @emw-apple does), they have presumably either passed `--no-add` to `git-webkit pr` or configured `webkitscmpy.auto-create-commit` to be `false`, both options which disable the automatic commit amending in this PR.

dewei-zhu commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
```diff
@@ -336,21 +336,42 @@ def pre_pr_checks(cls, repository):
+                    log.info('    Adding {}...'.format(file))
+                    if run([repository.executable(), 'add', file], cwd=repository.root_path).returncode:
+                        sys.stderr.write("Failed to add '{}'\n".format(file))
+                        return 1
```
What's the return value convention in this function? `True / False` for `pass / fail`, how about 1? Do you mean `True` here?

JonWBedard replied:
Good catch, this should be `return False`.

dewei-zhu reviewed (approved):
r=me with comment.

JonWBedard committed:
[git-webkit] Allow for re-running pre-pr checks
https://bugs.webkit.org/show_bug.cgi?id=242536
rdar://97093963

Reviewed by Dewei Zhu.

In webkit-patch, contributors could modify their patch after the style-checker
ran but before the change was uploaded. git-webkit should support a similar
functionality when running pre-pr checks.

* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.pre_pr_checks): If a pre-pr check fails, offer to re-run the
check and if possible, add any modified files to the most recent commit.
(PullRequest.create_pull_request): Pass --no-add onto pre_pr_checks.

Canonical link: https://commits.webkit.org/270007@main
